first_order_model/shrink_util.py

```diff
@@ -0,0 +1,1198 @@
+    """
+    model = copy.deepcopy(model)
+
+    """
+    Build layer_graph which graphs all the model dependencies
+    """
+    all_layers = [
+        m for n, m in model.named_modules()
+        if (isinstance(m, nn.Conv2d)
+            or isinstance(m, nn.modules.batchnorm._BatchNorm))
+    ]
+    layer_graph = build_graph(all_layers, [
+        n for n, m in model.named_modules()
+        if (isinstance(m, nn.Conv2d)
+            or isinstance(m, nn.modules.batchnorm._BatchNorm))
+    ])
+
+    # Deletions looks something like this
+    # {92: ['first', (94, 95)], 95: [(94, 95)], 93: [(94, 95)]}
+    for index, pruners in deletions.items():
+        node = layer_graph[index]
+
+        # A pruner marked with first means you delete from its outputs
+        # This can only happen for convolutional layers
+        if pruners[0] == 'first':
+            for prune_indices in reverse_sort(pruners[1:]):
+                if node.type == 'conv' and len(node.after) != 0:
+                    # Prune the outupts of the convolutional layer
+                    f_set(node.value.weight.detach(), node.value.weight, prune_indices)
+                    if node.value.bias is not None:
+                        f_set(node.value.bias.detach(), node.value.bias, prune_indices)
+                else:
+                    print("Should not be shrinking a batchnorm")
+                # Hacky solution to modify group count for dethwise convolutions
+                if node.value.groups != 1:
+                    node.value.groups = node.value.weight.shape[0]
+
+        else: # Delete from its inputs
+            for prune_indices in reverse_sort(pruners):
+
+                # Prune inputs of a convolutional layer
+                # We ignore depthwise layers since they take 1 input, their groups are just what changes.
+                if node.type == 'conv' and node.value.groups == 1:
+                    nvd = node.value.weight.detach()
+                    nvd = torch.cat(
+                        [nvd[:, :prune_indices[0]], nvd[:, prune_indices[1]:]],
+                        dim=1)
+                    node.value.weight.set_(nvd.clone().detach().contiguous())
+                
+                # Prune the inputs (and by definition outputs) of a batchnorm layer
+                if node.type == 'bn':
+
+                    f_set(node.value.weight.detach(), node.value.weight,
+                          prune_indices)
+                    f_set(node.value.bias.detach(), node.value.bias,
+                          prune_indices)
+                    f_set(node.value.running_mean.detach(),
+                          node.value.running_mean, prune_indices)
+                    f_set(node.value.running_var.detach(),
+                          node.value.running_var, prune_indices)
+
+    return model
+
+
+def get_metrics_loss(metrics_dataloader, lr_size, generator_full,
```

Comment:
if you have this function -- why don't you use it? Also doesn't seem relevant to `shrink_util` at all, and might as well use in `util.py`

Reply:
I do use it, it could be moved, but its also fine where it is since its only used in shrink_util.